File: dotnet-cleanroom/src/Botas/TurnContext.cs
using System.Text.Json;
using System.Net.Http.Json;

namespace Botas;

public class TurnContext
{
    private readonly BotApplication _app;
    private readonly CoreActivity _activity;

    public TurnContext(BotApplication app, CoreActivity activity)
    {
        _app = app;
        _activity = activity;
    }

    public CoreActivity Activity => _activity;
    public BotApplication App => _app;

    public async Task<ResourceResponse?> SendAsync(string text, CancellationToken ct = default)
    {
        var reply = new CoreActivity
        {
            Type = "message",
            Text = text,
            From = _activity.Recipient,
            Recipient = _activity.From,
            Conversation = _activity.Conversation
        };
        return await SendAsync(reply, ct);
    }

    public async Task<ResourceResponse?> SendAsync(CoreActivity activity, CancellationToken ct = default)
    {
        activity.ServiceUrl = _activity.ServiceUrl;
        activity.Conversation = _activity.Conversation;
        
        if (activity.From.Id == "")
            activity.From = _activity.Recipient;
        if (activity.Recipient.Id == "")
            activity.Recipient = _activity.From;

        return await _app.SendActivityAsync(activity, ct);
    }

    public async Task SendTypingAsync(CancellationToken ct = default)
    {
        var typing = new CoreActivity
        {
            Type = "typing",
            From = _activity.Recipient,
            Recipient = _activity.From,
            Conversation = _activity.Conversation,
            ServiceUrl = _activity.ServiceUrl
        };
        
        await _app.SendActivityAsync(typing, ct);
    }
}

File: python-cleanroom/packages/botas/src/turn_context.py
from typing import TYPE_CHECKING, Any, Union
from .models import CoreActivity, ResourceResponse

if TYPE_CHECKING:
    from .bot_application import BotApplication


class TurnContext:
    def __init__(self, app: "BotApplication", activity: CoreActivity) -> None:
        self._app = app
        self._activity = activity

    @property
    def activity(self) -> CoreActivity:
        return self._activity

    @property
    def app(self) -> "BotApplication":
        return self._app

    @property
    def from_account(self):
        return self._activity.from_account

    async def send(
        self, text_or_activity: Union[str, dict]
    ) -> ResourceResponse | None:
        if isinstance(text_or_activity, str):
            reply = CoreActivity(
                type="message",
                text=text_or_activity,
                from_account=self._activity.recipient,
                recipient=self._activity.from_account,
                conversation=self._activity.conversation,
            )
        else:
            reply = CoreActivity(**text_or_activity)
            reply.service_url = reply.service_url or self._activity.service_url
            reply.conversation = reply.conversation or self._activity.conversation
            reply.from_account = reply.from_account or self._activity.recipient
            reply.recipient = reply.recipient or self._activity.from_account

        return await self._app.send_activity_async(
            self._activity.service_url, self._activity.conversation.id, reply
        )

    async def send_typing(self) -> None:
        typing = CoreActivity(
            type="typing",
            from_account=self._activity.recipient,
            recipient=self._activity.from_account,
            conversation=self._activity.conversation,
        )
        await self._app.send_activity_async(
            self._activity.service_url, self._activity.conversation.id, typing
        )